### ISSUE TITLE
Incorrect annotation on JDK Constructor.newInstance

### ISSUE THREAD

**Issue by diamondq** [typetools/checker-framework#1365]
In the annotated JDK (v2.1.12), the `java.lang.reflect.Constructor.newInstance` should support passing a null array. However, it doesn't currently. This should work:
```java
java.lang.reflect.Constructor c = String.class.getConstructor(String.class);
Object[] p = null;
c.newInstance(p);
```
but the Nullness Checker returns with a:

```
Error: [argument.type.incompatible] incompatible types in argument.
found : @Initialized @NonNull Object @FBCBottom @Nullable []
required: @Initialized @Nullable Object @Initialized @NonNull []
```
It looks like the JDK was annotated to support a Nullable element, but not a Nullable array (see #336).

Cheers

**Comment by wmdietlGC** [comment]
Thanks for the report!

The javadoc at https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Constructor.html#newInstance(java.lang.Object...) states:
`If the number of formal parameters required by the underlying constructor is 0, the supplied initargs array may be of length 0 or null.`

We should therefore change the annotations to:

```java
    public @NonNull T newInstance(@Nullable Object @Nullable ... initargs) throws InstantiationException,IllegalAccessException,IllegalArgumentException,InvocationTargetException { throw new RuntimeException("skeleton method"); }
```

I'll prepare a PR later tonight.

**Comment by mernst** [comment]
I believe that the current annotation is correct, so no change should be made.

If a formal parameter is annotated as `@Nullable`, that means that `null` is *always* a legal argument.
In the case of `getConstructor`, `null` is sometimes legal and sometimes illegal.  If the Nullness Checker always permitted `null` to be passed to `getConstructor`, then it would permit some illegal calls to `getConstructor`.  Forbidding some legal calls is the price of forbidding all illegal calls.  That is, some false positives are the price of verification.

It would be possible to extend the Nullness Checker to treat `getConstructor` more precisely.  The Nullness Checker would need to analyze which constructor is being called (it could use the Constant Value Checker to do this); when the constructor is known to require no formal parameters, the Nullness Checker would permit the argument to be null.

**Comment by wmdietlGC** [comment]
@mernst The current annotation for `Constructor.newInstance` is:

```java
    public @NonNull T newInstance(@Nullable Object ... initargs) throws InstantiationException,IllegalAccessException,IllegalArgumentException,InvocationTargetException { throw new RuntimeException("skeleton method"); }
```

This is neither completely conservative nor fully flexible: it allows null arguments, which is also not always legal.
So the current annotations are definitely not correct.
To be fully conservative, we would need to change the signature to:

```java
    public @NonNull T newInstance(Object ... initargs) throws InstantiationException,IllegalAccessException,IllegalArgumentException,InvocationTargetException { throw new RuntimeException("skeleton method"); }
```

Note that the signature for Method.invoke is:

```java
    public @Nullable Object invoke(Object obj, Object ... args)
```

where neither the elements nor the array itself are `@Nullable`.

So to be consistent, let's remove `@Nullable` from the component type of `Constructor.newInstance`.

Also note https://checkerframework.org/manual/#reflection-resolution which implements the reflection resolution and would reduce the conservative warnings, if it can prove the concrete types.

**Comment by mernst** [comment]
OK, that sounds good.  (And thanks correcting the details!)
@diamondq, would reflection resolution help you in this case?

**Comment by diamondq** [comment]
Sorry, I guess I'm a little confused. I think that I follow the conversation, but if you remove all the `@Nullable` from the parameters, then how do I call a `Constructor.newInstance` (or for that matter, an `Method.invoke`) and pass a null value without generating an warning/error?

I can sort of work around the `newInstance` case by passing a `new Object[0]`, but I don't see how I can work around the `Method.invoke` when I'm calling a static method. The first parameter is MUST be null to call a static method. 

I realize that `Method.invoke` is a separate issue from the `newInstance` that this issue was created around, but I'm trying to understand the general thought patterns.

**Comment by wmdietl** [comment]
For an arbitrary `Method.invoke` you don't know whether the receiver is allowed to be null or not - the analyzer needs to know whether the method is static or not. Similarly for the parameter and return types. The Checker Framework wants to prevent all `NullPointerException`s and if the checker doesn't know whether the method is static, it has to warn about the null argument for the receiver. If the method is an instance method, passing null as the receiver will give an NPE.
As @mernst put it: "Forbidding some legal calls is the price of forbidding all illegal calls. That is, some false positives are the price of verification."

If reflection resolution doesn't solve this issue, but you have some additional, application-specific logic why the invocation is valid, e.g. why the method being invoked is always static, you can suppress the warning and document your assumptions.

**Comment by diamondq** [comment]
Unfortunately, the reflection resolution doesn't appear to work in my case, as the definition of the Constructor happens elsewhere, and presumably the Checker can't determine what the Constructor is pointing to.

I'll just go with a `@SuppressWarnings("nullness")` and try to keep the suppression in as small a fragment of code as possible.

Thanks for discussion. I can totally understand the balancing act of trying to minimize false positives while focusing on removing all false negatives.

Feel free to close (or keep it open to track the adjustment to the Constructor)
